examples/recurrent_attention_machine_translation.py

```diff
@@ -0,0 +1,961 @@
+
+        if t_max is None:
+            t_max = x_.shape[-1] * 2
+
+        start_idx = target_tokenizer.word_index[start_token]
+        end_idx = target_tokenizer.word_index[end_token]
+
+        # A list of the <search_width> number of beams with highest score is
+        # maintained through out the search. Initially there is only one beam.
+        incomplete_beams = [Beam(sequence=[start_idx], scores=[0])]
+        # All beams that reached the end-token are kept separately.
+        complete_beams = []
+
+        t = 0
+        while len(complete_beams) < search_width and t < t_max:
+            t += 1
+            # create a batch of inputs representing the incomplete_beams
+            y_ = np.vstack([beam.sequence for beam in incomplete_beams])
+            # predict next tokes for every incomplete beam
+            batch_size = len(incomplete_beams)
+            y_pred_ = model.predict([y_, x_[:batch_size]])
+            # from each previous beam create new candidate beams and save the once
+            # with highest score for next iteration.
+            beams_updated = []
+            for i, beam in enumerate(incomplete_beams):
+                for proba, idx in k_largest_val_idx(y_pred_[i, -1], branch_factor):
+                    new_beam = beam.get_child(element=idx, score=np.log(proba))
+                    if idx == end_idx:
+                        # beam completed
+                        complete_beams.append(new_beam)
+                    elif len(beams_updated) < search_width:
+                        # not full search width
+                        heapq.heappush(beams_updated, new_beam)
+                    elif new_beam > beams_updated[0]:
+                        # better than candidate with lowest score
+                        heapq.heapreplace(beams_updated, new_beam)
+                    else:
+                        # if score is worse that existing candidates we abort search
+                        # for children of this beam (since next token processed
+                        # in order of decreasing likelihood)
+                        break
+
+            # faster to process beams in order of decreasing score next iteration,
+            # due to break above
+            incomplete_beams = sorted(beams_updated, reverse=True)
+
+        # want to return in order of decreasing score
+        complete_beams = sorted(complete_beams, reverse=True)
+
+        output_texts = []
+        output_scores = []
+        for beam in complete_beams + incomplete_beams:
+            text = target_tokenizer.sequences_to_texts(beam.sequence[None])[0]
+            output_texts.append(text)
+            # average score, skipping start token
+            output_scores.append(beam.score / (len(beam) - 1))
+
+        return output_texts, output_scores
+
+    # Translate first 3 samples from validation data
+    for input_text, target_text in zip(input_texts_val, target_texts_val)[:3]:
```

Comment:
Change to:

for input_text, target_text in list(zip(input_texts_val, target_texts_val))[:3]:

zip is eager in PY2, lazy in PY3